### ISSUE TITLE
[TCGC] propertyFlatten not apply to property of polymorphic type

### ISSUE THREAD

**Issue by weidongxu-microsoft** [Azure/typespec-azure#959]
### Clear and concise description of the problem

Context: at present, `@flattenProperty` had to be applied to template property. e.g. https://github.com/Azure/azure-rest-api-specs/blob/1418642cf8a57954c781c26c759fd4aef39a1287/specification/vmware/Microsoft.AVS/client.tsp#L19

For a property of polymorphic type, it is not possible to flatten the model, as polymorphic is runtime behavior.
https://github.com/Azure/azure-rest-api-specs/blob/1418642cf8a57954c781c26c759fd4aef39a1287/specification/vmware/Microsoft.AVS/models.tsp#L1715-L1744

Therefore, either the decorator, or the TCGC, may want to not set `SdkBodyModelPropertyType.flatten=true`.

Alternatively, this can be done in language emitter part (Java does this), as e.g. one would still encounter it from Swagger. However, the flatten may already be done by m4.
https://github.com/Azure/azure-rest-api-specs/blob/32f566eb38af9593cfc1ed183362471c50d67fb2/specification/vmware/resource-manager/Microsoft.AVS/stable/2023-09-01/vmware.json#L6514-L6516

### Checklist

- [X] Follow our [Code of Conduct](https://github.com/azure/typespec-azure/blob/main/CODE_OF_CONDUCT.md)
- [X] Check that this issue is about the Azure libraries for typespec. For feature request in the typespec language or core libraries file it in the [TypeSpec repo](https://github.com/Microsoft/TypeSpec/issues/new/choose)
- [X] Read the [docs](https://azure.github.io/typespec-azure/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by tadelesh** [comment]
@archerzz  i believe the rule could  be added in `@flattenProperty` decorator function, to judge whether the property type is a base discriminated model.

**Comment by archerzz** [comment]
A couple of issues:
1. Why there is no error when `Azure.ResourceManager.ProxyResource.properties` is the target? There seems some problem in TypeSpec compiler when applying decorator to target. See here: https://github.com/microsoft/typespec/issues/3406
2. This could be a feature in TypeSpec that allowing people to assign a decorator to template properties, like
```typescript
model SomeTemplate<T> {
    @flattenProperty
    properties: T
}
```
3. If we think $2 above is not supported, then we need to throw error if a decorator is used on template property. But we need to fix issue $1 first

For a work-around, we should tell service team to decorate concrete models, like:
```typescript
@@flattenProperty(Addon .properties);
@@flattenProperty(Location.properties);
```

**Comment by archerzz** [comment]
For $3, the corresponding issue is targeted for `[2024] - July`

**Comment by tadelesh** [comment]
@archerzz  this issue is for disallowing `@propertyFlatten` on polymorphic types, not for the issue you mentioned.

**Comment by archerzz** [comment]
fixed in #1223
